File: Leetcode/04_Strings/027_0459_repeated-substring-pattern.go
// https://leetcode.cn/problems/repeated-substring-pattern/description/
// e

package main

import "strings"

// 把给的字符串*2, 扣除首尾第一个字符如果还包含给的字符串本身，则返回 true
func repeatedSubstringPattern(s string) bool {
	return strings.Contains((s + s)[1:len(s)*2-1], s)
}

// KMP算法更加高效
func repeatedSubstringPattern_KMP(s string) bool {
	// 首先算出 next
	n := len(s)
	next := make([]int, len(s))
	i := 0
	for j := 1; j < n; j++ {
		for i > 0 && s[i] != s[j] {
			i = next[i-1]
		}
		if s[i] == s[j] {
			i++
		}
		next[j] = i
	}

	// 然后判断 len(s) 能否被 len(s) - next[-1] 整除
	return next[n-1] > 0 && n%(n-next[n-1]) == 0
}

File: Leetcode/04_Strings/027_0459_repeated-substring-pattern.py
# https://leetcode.cn/problems/repeated-substring-pattern/description/
# e

# 把给的字符串*2, 扣除首尾第一个字符如果还包含给的字符串本身，则返回 true
class Solution(object):
    def repeatedSubstringPattern(self, s):
        ns = s + s
        n = len(s)
        for i in range(1, n):
            if s == ns[i:i+n]:
                return True
        return False
    
class Solution(object):
    def repeatedSubstringPattern(self, s):
        return s in (s*2)[1:-1]


# KMP解法，更高效：
class Solution(object):
    def repeatedSubstringPattern(self, s):
        # KMP 算法求出 next
        n = len(s)
        nxt = [0] * n
        i = 0
        for j in range(1, n):
            while s[j] != s[i] and i > 0:
                i = nxt[i-1]
            if s[j] == s[i]:
                i += 1
            nxt[j] = i
        
        # 判断能否整除
        rem = n - nxt[n-1]
        return nxt[n-1] > 0 and n % rem == 0